tests/net/conn_mgr/src/main.c

```diff
@@ -0,0 +1,921 @@
+	zassert_equal(stats.event_count, 0,
+		"No events should be fired if connectivity availability did not change.");
+
+	/* Take A down */
+	if (conn_mgr_if_is_bound(ifa)) {
+		zassert_equal(conn_mgr_if_disconnect(ifa), 0,
+			"conn_mgr_if_disconnect should succeed for ifa.");
+	}
+	zassert_equal(net_if_down(ifa), 0, "net_if_down should succeed for ifa.");
+
+	/* Expect connectivity lost */
+	k_sleep(EVENT_WAIT_TIME);
+	stats = get_reset_stats();
+	zassert_equal(stats.dconn_count, 1,
+		"NET_EVENT_L4_DISCONNECTED should be fired when connectivity is lost.");
+	zassert_equal(stats.event_count, 1,
+		"Only NET_EVENT_L4_DISCONNECTED should be fired when connectivity is lost.");
+	zassert_equal(stats.dconn_iface, ifa, "ifa should be blamed.");
+
+	/* Take B down */
+	if (conn_mgr_if_is_bound(ifb)) {
+		zassert_equal(conn_mgr_if_disconnect(ifb), 0,
+			"conn_mgr_if_disconnect should succeed for ifb.");
+	}
+	zassert_equal(net_if_down(ifb), 0, "net_if_down should succeed for ifb.");
+
+	/* Expect no events */
+	k_sleep(EVENT_WAIT_TIME);
+	stats = get_reset_stats();
+	zassert_equal(stats.event_count, 0,
+		"No events should be fired if connectivity availability did not change.");
+}
+
+enum ip_order {
+	IPV4_FIRST,
+	IPV6_FIRST
+};
+
+/**
+ * @brief Cycles a single iface through all possible ready and pre-ready states,
+ * ensuring the correct events are observed and generated by conn_mgr.
+ *
+ * Ifaces can be in one of four states that are relevant to L4 readiness:
+ * 00: oper-down, no IPs associated (unready state)
+ * 01: Has IP, is oper-down (semi-ready state)
+ * 10: Is oper-up, has no IP (semi-ready state)
+ * 11: Has IP and is oper-up (ready state)
+ *
+ * In total there are eight possible state transitions:
+ *
+ * (00 -> 10): Gain oper-up from unready state
+ * (10 -> 11): Gain IP from semi-ready state
+ * (11 -> 10): Lose IP from ready state
+ * (10 -> 00): Lose oper-up from semi-ready state
+ * (00 -> 01): Gain IP from unready state
+ * (01 -> 11): Gain Oper-up from semi-ready state
+ * (11 -> 01): Lose oper-up from ready state
+ * (01 -> 00): Lose IP from semi-ready state
+ *
+ * We test these state transitions in that order.
+ *
+ * This is slightly complicated by the fact that ifaces can be assigned multiple IPs, and multiple
+ * types of IPs. Whenever IPs are assigned or removed, two of them, an IPv6 and IPv4 address is
+ * added or removed.
+ *
+ * @param iface
+ * @param ifa_ipm
+ */
+static void cycle_iface_states(struct net_if *iface, enum ip_order ifa_ipm)
+{
+	struct test_stats stats;
+
+	/* DAD messes up the IP addition/removal logic of this test routine, and it is sufficient
+	 * to execute it only when DAD is disabled, since we don't expect DAD to affect any
+	 * behaviors beyond the IP addition/removal testing it defeats.
+	 *
+	 * Compatibility of conn_mgr with DAD is ensured by cycle_ignored_iface succeeding both
+	 * with and without DAD (the suite is configured to run twice, once with each setting).
```

Comment:
Thinking on it, this suite currently lacks a test explicitly requiring DAD to succeed as part of the test. I will write one.